File: qr/expo/app/modules/bench-marker/ios/BenchMarkerModule.swift
import ExpoModulesCore
import Foundation
import os

/// Writes the benchmark markers described in AGENTS.md to the platform log.
///
/// JavaScript takes the timestamps and hands them over through this module;
/// only `processStart` is taken here, because JavaScript cannot see it.
/// `console.log` is not used: it does not reach the app log in a release build.
public class BenchMarkerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("BenchMarker")

    // Synchronous, so reporting a marker cannot be deferred past the moment
    // it is measuring.
    Function("mark") { (name: String, epochMs: Double) in
      BenchMarker.mark(name, epochMs: Int64(epochMs))
    }

    /// Emits `processStart` as well, the one marker JavaScript cannot know.
    Function("markLaunch") { (name: String, epochMs: Double) in
      if let start = BenchMarker.processStartEpochMs() {
        BenchMarker.mark("processStart", epochMs: start)
      }
      BenchMarker.mark(name, epochMs: Int64(epochMs))
    }

    // A figure rather than a time; see AGENTS.md.
    Function("markValue") { (name: String, value: String) in
      BenchMarker.markValue(name, value: value)
    }

    // The benchmark's preparation step, which is not measured: unpack the
    // bundled QR images into the app's own storage so the decode loop can
    // read them by file path, like the other implementations do.
    Function("prepareImages") { () -> [String] in
      try QrImages.unpack()
    }
  }
}

enum BenchMarker {
  private static let logger = Logger(subsystem: "bench", category: "marker")

  static func mark(_ name: String, epochMs: Int64) {
    // Interpolated values are redacted as <private> in release builds unless
    // they are explicitly public.
    logger.notice("BENCH|\(name, privacy: .public)|\(epochMs, privacy: .public)")
    // On a physical iPhone the app log carries only the process's own output,
    // not unified logging, so the line goes to stderr as well.
    fputs("BENCH|\(name)|\(epochMs)\n", stderr)
  }

  static func markValue(_ name: String, value: String) {
    logger.notice("BENCHVAL|\(name, privacy: .public)|\(value, privacy: .public)")
    fputs("BENCHVAL|\(name)|\(value)\n", stderr)
  }

  /// When the kernel started this process, which is where a cold start begins.
  static func processStartEpochMs() -> Int64? {
    var info = kinfo_proc()
    var size = MemoryLayout<kinfo_proc>.stride
    var mib: [Int32] = [CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()]
    guard sysctl(&mib, u_int(mib.count), &info, &size, nil, 0) == 0 else {
      return nil
    }
    let start = info.kp_proc.p_un.__p_starttime
    let seconds = TimeInterval(start.tv_sec) + TimeInterval(start.tv_usec) / 1_000_000
    return Int64((seconds * 1000).rounded())
  }
}

enum QrImages {
  /// Copies `qr/*.png` out of the app bundle into Documents, and returns the
  /// paths in name order. Images already there are left alone.
  static func unpack() throws -> [String] {
    guard let source = Bundle.main.resourceURL?.appendingPathComponent("qr") else {
      throw NSError(domain: "bench", code: 1, userInfo: [NSLocalizedDescriptionKey: "no bundle"])
    }
    let documents = try FileManager.default.url(
      for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
    let destination = documents.appendingPathComponent("qr")
    try FileManager.default.createDirectory(at: destination, withIntermediateDirectories: true)

    let names = try FileManager.default.contentsOfDirectory(atPath: source.path)
      .filter { $0.hasSuffix(".png") }
      .sorted()
    return try names.map { name in
      let file = destination.appendingPathComponent(name)
      // An empty file means a half-finished copy from an earlier launch.
      let size = (try? file.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0
      if size == 0 {
        try? FileManager.default.removeItem(at: file)
        try FileManager.default.copyItem(at: source.appendingPathComponent(name), to: file)
      }
      return file.path
    }
  }
}

File: qr/expo/app/modules/bench-marker/android/src/main/java/expo/modules/benchmarker/BenchMarkerModule.kt
package expo.modules.benchmarker

import android.os.Process
import android.os.SystemClock
import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

/**
 * Writes the benchmark markers described in AGENTS.md to logcat.
 *
 * JavaScript takes the timestamps and hands them over through this module;
 * only `processStart` is taken here, because JavaScript cannot see it.
 * `console.log` is not used: it does not reach logcat in a release build.
 */
class BenchMarkerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("BenchMarker")

    // Synchronous, so reporting a marker cannot be deferred past the moment
    // it is measuring.
    Function("mark") { name: String, epochMs: Double -> mark(name, epochMs.toLong()) }

    // Emits `processStart` as well, the one marker JavaScript cannot know.
    Function("markLaunch") { name: String, epochMs: Double ->
      mark("processStart", processStartEpochMs())
      mark(name, epochMs.toLong())
    }

    // A figure rather than a time; see AGENTS.md.
    Function("markValue") { name: String, value: String -> markValue(name, value) }

    // The benchmark's preparation step, which is not measured: unpack the
    // bundled QR images into the app's own storage so the decode loop can
    // read them by file path, like the other implementations do.
    Function("prepareImages") { unpackImages() }
  }

  /**
   * Copies the PNGs under `assets/qr` out of the APK into filesDir, and
   * returns the paths in name order. Images already there are left alone.
   * An APK asset has no file path of its own, which is why this exists.
   *
   * (A KDoc cannot say `qr` followed by a slash and a star: Kotlin nests
   * block comments, so that would open one.)
   */
  private fun unpackImages(): List<String> {
    val context = appContext.reactContext ?: error("no context")
    val destination = File(context.filesDir, "qr")
    destination.mkdirs()
    val names = (context.assets.list("qr") ?: emptyArray()).filter { it.endsWith(".png") }.sorted()
    return names.map { name ->
      val file = File(destination, name)
      // An empty file means a half-finished copy from an earlier launch.
      if (!file.exists() || file.length() == 0L) {
        context.assets.open("qr/$name").use { input ->
          file.outputStream().use { output -> input.copyTo(output) }
        }
      }
      file.absolutePath
    }
  }

  private fun markValue(name: String, value: String) {
    Log.i("Bench", "BENCHVAL|$name|$value")
  }

  private fun mark(name: String, epochMs: Long) {
    Log.i("Bench", "BENCH|$name|$epochMs")
  }

  /** When this process started, converted from uptime to wall-clock time. */
  private fun processStartEpochMs(): Long =
    System.currentTimeMillis() - (SystemClock.uptimeMillis() - Process.getStartUptimeMillis())
}
